stmtools/stm.py

```diff
@@ -569,6 +669,82 @@ def _compute_morton_code(xx, yy):
     -------
     array_like
         An array with Morton codes per coordinate pair.
+
     """
     code = [pm.interleave(int(xi), int(yi)) for xi, yi in zip(xx, yy, strict=True)]
     return code
+
+
+def _enrich_from_raster_block(ds, dataraster, fields, method):
+    """Enrich the ds (SpaceTimeMatrix) from one or more fields of a raster dataset.
+
+    scipy is required. It uses xarray.Dataset.interp_like to interpolate the
+    raster dataset to the coordinates of ds.
+    https://docs.xarray.dev/en/stable/generated/xarray.Dataset.interp.html
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        SpaceTimeMatrix to enrich
+    dataraster : xarray.Dataset | xarray.DataArray
+        Input data for enrichment
+    fields : str or list of str
+        Field name(s) in the dataset for enrichment
+    method : str, optional
+        Method of interpolation, by default "nearest", see
+
+    Returns
+    -------
+    xarray.Dataset
+
+    """
+    # interpolate the raster dataset to the coordinates of ds
+    interpolated = dataraster.interp(ds.coords, method=method)
+
+    # Assign these values to the corresponding points in ds
+    for field in fields:
+        ds[field] = xr.DataArray(interpolated[field].data, dims=ds.dims, coords=ds.coords)
+    return ds
+
+
+def _enrich_from_points_block(ds, datapoints, fields):
+    """Enrich the ds (SpaceTimeMatrix) from one or more fields of a point dataset.
+
+    https://docs.xarray.dev/en/latest/generated/xarray.DataArray.sel.html#xarray.DataArray.sel
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        SpaceTimeMatrix to enrich
+    datapoints : xarray.Dataset | xarray.DataArray
+        Input data for enrichment
+    fields : str or list of str
+        Field name(s) in the dataset for enrichment
+
+    Returns
+    -------
+    xarray.Dataset
+
+    """
+    # unstak the dimensions
+    for dim in datapoints.dims:
+        if dim not in datapoints.coords:
+            indexer = {
+                dim: [
+                    coord for coord in datapoints.coords if dim in datapoints[coord].dims
+                    ]
+                }
+            datapoints = datapoints.set_index(indexer)
+            datapoints = datapoints.unstack(dim)
```

Comment:
`unstack` operation will evaluate the rasterized datapoints. When `datapoints` is large, e.g. it has a scope which is much larger than `ds`, this can cause memory issues. Since this is a block-wise function, I would suggest adding bounding box selection at the beginning of `_enrich_from_points_block`.

Reply:
My understanding is that if the Dataset or variable is chunked, then it's a dask array and computations on it are lazy until the result is needed by calling `.values `or `.compute()` on the Dataset.  To check this, I wrote tests like `test_all_operations_lazy`. It is also possible to check this by adding `print(datapoints.chunks)` or `print(datapoints.temperature)` whether it returns a `dask.array`. Can you please explain how "`unstack` operation will evaluate the rasterized datapoints." ?

Reply:
I replaced the approach of using xarray `set_index` and `unstack` with `scipy.spatial.KDTRee`. It is faster with a large dataset. So, your example test in the notebook is working now. The reason is that `unstack` does not work with dask array. Also functions like `map_blocks` are not helpful because unstack changes the shape of the dataset.